CIP-0133/README.md

```diff
@@ -0,0 +1,116 @@
+- **Mithril** - a protocol for helping to scale the adoption of Cardano and its accessibility for users. It creates certified snapshots of the Cardano blockchain allowing to obtain a verified version of the current state without having to download and verify the full history of the blockchain. Mithril utilizes a stake-based threshold multisignature scheme based on elliptic curve pairings. Even though at the moment most use cases of Mithril relies on off-chain computations, eventually the Mithril certificates might also be verified in Plutus smart contracts.
+- **ATALA** - a decentralized identification mechanism. One of the properties it can provide is anonymity: users can selectively disclose attributes of their certificate or prove statements without disclosing their identity. Up to date, the most efficient solutions for doing that use pairing-based zero-knowledge protocols.
+
+In conclusion, integrating multi-scalar multiplication as a core built-in within Plutus is not only essential for enhancing cryptographic capabilities, but also for optimizing on-chain computations. The current approach of naive manual implementation in Plutus is inefficient for large-scale MSMs. Incorporating this function directly will streamline these operations, reduce transaction costs, and maintain the integrity of existing tools, thereby significantly advancing the Plutus ecosystem's functionality and user experience.
+
+## Specification
+The MSM for BLS12-381 is implemented in [blst](https://github.com/supranational/blst/blob/e99f7db0db413e2efefcfd077a4e335766f39c27/bindings/blst.h#L241) library, which is already a dependency for the [cardano-base](https://github.com/IntersectMBO/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/). This library provides efficient and formally verified implementation of operations over the BLS12-381 elliptic curve. It has been used for implementing [CIP-0381](https://cips.cardano.org/cip/CIP-0381). Basically, we would like to expose several additional functions from this library in Plutus API.
+
+### Function definition
+We propose to define two new Plutus functions **bls12_381_G1_multiScalarMul** and **bls12_381_G2_multiScalarMul** as follows:
+
+```hs
+bls12_381_G1_multiScalarMul :: [Integer] -> [bls12_381_G1_element] -> bls12_381_G1_element
+bls12_381_G2_multiScalarMul :: [Integer] -> [bls12_381_G2_element] -> bls12_381_G2_element
+```
+
+The types **bls12_381_G1_element** and **bls12_381_G2_element** are already introduced by [CIP-0381](https://cips.cardano.org/cip/CIP-0381). Given two arrays of scalars and group elements the functions compute multi-scalar multiplication for the corresponding subgroup. The arrays of scalars and group elements must be non-empty and have equal size.
```

Comment:
Rust bindings in blst truncate the longer list and panic if the size is zero. As far as I can see pure C bindings doesn't handle empty list either.
I'm not that familiar with Plutus and how it handles failures, but I think it's better to return some sort of an error.

Reply:
@kwxm, I think this is the same problem we had in [this](https://github.com/input-output-hk/cardano-base/pull/289) fix? In that, there is some assumption on the structure of the data the lib expects, and that the bindings should check for those. We should indeed make sure we match those to prevent unexpected failures.

Reply:
@dkaidalov, do you know why the rust bindings chose to truncate the bigger list of the two? Why not fail if they are not equal length? We specified that they should be of equal length.

Reply:
ah, no, actually rust bindings in [blst](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger.rs#L138) panic if lists are not equal. 
It's another library on top of blst that we used for benchmarks [does](https://github.com/filecoin-project/blstrs/blob/master/src/g1.rs#L640) truncation of scalars but still fails if the list of points is larger. I don't know why.

Reply:
The CIP should probably say exactly what behaviour's expected in the case of empty lists (return the identity or return `Nothing` or fail in some other way?) and lists of non-equal length (truncate or return `Nothing` or fail in some other way?).  The  Haskell interface can deal with these situations, so it doesn't really matter what the C library does; in fact it'd probably be easier to check these cases in the Haskell code rather than trying to intercept an error from C.